jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletApiRequest.java

```diff
@@ -734,8 +737,68 @@ public Part getPart(String name) throws IOException, ServletException
     @Override
     public <T extends HttpUpgradeHandler> T upgrade(Class<T> handlerClass) throws IOException, ServletException
     {
-        // Not implemented. Throw ServletException as per spec.
-        throw new ServletException("Not implemented");
+        Response response = _servletContextRequest.getServletContextResponse();
+        if (response.getStatus() != HttpStatus.SWITCHING_PROTOCOLS_101)
+            throw new IllegalStateException("Response status should be 101");
+        if (response.getHeaders().get("Upgrade") == null)
+            throw new IllegalStateException("Missing Upgrade header");
+        if (!"Upgrade".equalsIgnoreCase(response.getHeaders().get("Connection")))
+            throw new IllegalStateException("Invalid Connection header");
+        if (response.isCommitted())
+            throw new IllegalStateException("Cannot upgrade committed response");
+        if (_servletChannel.getConnectionMetaData().getHttpVersion() != HttpVersion.HTTP_1_1)
+            throw new IllegalStateException("Only requests over HTTP/1.1 can be upgraded");
+
+        ServletOutputStream outputStream = _servletContextRequest.getHttpOutput();
+        ServletInputStream inputStream = _servletContextRequest.getHttpInput();
+
+        T upgradeHandler;
+        try
+        {
+            upgradeHandler = handlerClass.getDeclaredConstructor().newInstance();
+        }
+        catch (Exception e)
+        {
+            throw new ServletException("Unable to instantiate handler class", e);
+        }
+
+        HttpConnection httpConnection = (HttpConnection)_servletContextRequest.getConnectionMetaData().getConnection();
+        httpConnection.getParser().servletUpgrade();
+        AsyncContext asyncContext = forceStartAsync(); // force the servlet in async mode
+
+        outputStream.flush(); // commit the 101 response
+        httpConnection.getGenerator().servletUpgrade(); // tell the generator it can send data as-is
+
+        WebConnection webConnection = new WebConnection()
+        {
+            @Override
+            public void close() throws Exception
+            {
+                IO.close(inputStream);
+                IO.close(outputStream);
+            }
+
+            @Override
+            public ServletInputStream getInputStream()
+            {
+                return inputStream;
+            }
+
+            @Override
+            public ServletOutputStream getOutputStream()
+            {
+                return outputStream;
+            }
+        };
+
+        _servletContextRequest.getServletChannel().setCompleteListener(() ->
+        {
+            upgradeHandler.destroy();
+            asyncContext.complete();
+        });
```

Comment:
I don't think this mechanism should be on ServletChannel.  This method already has access to both the `HttpInput` and `HttpOutput` classes, so at the very least it could inject something to them, or better yet just wrap them somehow.

Eitherway, let's lose the ServletChannel listener mechanism